### DISCUSSION TITLE
Custom errors management with partial results

### DISCUSSION THREAD

**Discussion by tvallois** [mirumee/ariadne#1016] [Q&A] [answered]
Hello,

I don't know if the question has already been asked/answered, i don't find any traces of it.
I try to follow [GraphQL specs](https://spec.graphql.org/draft/#sec-Errors) to manage errors with our application.
Imagine a scenario where i try to create 3 tokens linked to emails using a mutation:
```graphql
type Mutation {
    createTokens(
         emails: [String!]!
    ): CreateTokensPayload!
}

type Token {
   data: String!
   email: String!
}

type CreateTokensPayload {
    tokens: [Token]
}
```

If everything work fine, i should have this result:
```json
{
  "data": {
    "createTokens": [
      {
        "data": "sze",
        "email": "email1@gmail.com"
      },
      {
        "data": "azead",
        "email": "email2@gmail.com"
      },
      {
        "data": "adzadzd",
        "email": "email3@gmail.com"
      }
    ]
  }
}
```

If an error occur only on email `email2@gmail.com`, according to GraphQL specs, i should have this kind of result:
```json
{
  "errors": [
    {
      "message": "error message",
      "locations": [
        {
          "line": 6,
          "column": 7
        }
      ],
      "path": [
        "createTokens",
        1,
        "email"
      ]
    }
  ],
  "data": {
    "createTokens": [
      {
        "data": "sze",
        "email": "email1@gmail.com"
      },
      null,
      {
        "data": "adzadzd",
        "email": "email3@gmail.com"
      }
    ]
  }
}
```

I don't find a way to do this with Ariadne, any suggestions?

Thanks!

**Comment by rafalp** [comment, marked as answer]
The part of errors spec you are referring to is about errors during data retrieval. This means that error for invalid e-mail should be raised in resolver for `Token`'s `email` field.

For example you `createTokens` mutation could return this data:

```python
{
    "tokens": [
        {"data": "valid data", "email": "valid@example.com", "error": None},
        {"data": "valid data", "email": None, "error": "Invalid email!"},
        {"data": "valid data", "email": "valid2@example.com", "error": None},
    ]
}
```

With email resolver like this:

```python
def resolve_email(obj, *_):
    if obj.get("error"):
        raise ValueError(obj["error"])

    return obj["email"]
```

**Reply by tvallois** [reply]
Hmm, it was the first implementation i did but i'm not sure this is the way to do it.
Looking at [Apollo server implementation](https://www.apollographql.com/docs/apollo-server/data/errors), `errors` seems separated from `data` result.

**Reply by rafalp** [reply]
GraphQL spec only says that errors occurred during request handling will be returned by the server under `errors` key which is separated from `data`. It doesn't say a word about how server implementation should move errors around internally, so its perfectly reasonable for resolver to have error information as part of data it passes to child errors.

However it is considered a bad practice to use GraphQL's standard error mechanic for error reporting, exactly for the reason that it needs extra complexity to work around spec's. Moving error down few fields so its raised from other field personally feels like a hack to me. Those errors are also non-typed, so you can't learn what errors to expect from GraphQL by looking at it's schema. Its generally much easier for both server and client developers to work with errors that are part of schema.

We even mention that in [Ariadne's documentation](https://ariadnegraphql.org/docs/error-messaging).

**Reply by tvallois** [reply]
Ok. To be sure to understand, it's a design choice by default for Ariadne which is different from [Apollo design](https://www.apollographql.com/docs/react/data/error-handling/) right?

**Reply by rafalp** [reply]
No, it's not. Both Ariadne and Apollo behave the same in this field. But Apollo documentation suggests you to extend the spec for reporting errors specific to the application, and Ariadne tells you to define custom error types that are part of schema for the same.

 I'll explain this in more detail now, but key difference between what you are trying to accomplish and what Apollo shows, is that you are trying to return error for a sub field in GraphQL type resulted by **mutation**, and their example returns an error from top level field in **query**.

In their example field they want to return error for is the same field that that implements logic, the `userWithID` field. In your example logic lives in the `createTokens` field, but you want those errors come from  individual `email` fields belonging to the `Token` type which is returned from `tokens` field belonging to `CreateTokensPayload`.

So you need to first pass your error down to the `email` field before it can be raised. If you were using Apollo server (or any other spec compliant server), it will be exactly the same. You would have to pass the information about the error down to an email field.

- - - - -

I've just remembered that part of GraphQL spec is that if resolver field returns error (not raises but returns), query executor should treat this situation the same to error being raised from resolver. So you could skip custom `email` resolver and return data from `createTokens` that looks like this:

```python
{
    "tokens": [
        {"data": "valid data", "email": "valid@example.com", "error": None},
        {"data": "valid data", "email": ValueError("Invalid email!")},
        {"data": "valid data", "email": "valid2@example.com", "error": None},
    ]
}
```

**Reply by tvallois** [reply]
I'm not sure to fully understand your explanation.
I tried to do something like this:
```python
from ariadne import MutationType

def create_tokens(*_, emails):
    tokens = [{"token": "fsqd", "email": ValueError("boom"), {"token": "aze", "email": emails[0]}}]
    return invalid_tokens

mutation_type = MutationType()
mutation_type.set_field("createTokens", create_tokens)
```

This solution return:
```
{'data': {'createTokens': {'tokens': None}}, 'errors': [{'message': 'boom', 'locations': [{'line': 14, 'column': 21}], 'path': ['createTokens', 'tokens', 0, 'email']}]}
```

**Reply by rafalp** [reply]
```python
tokens = [{"token": "fsqd", "email": ValueError("boom"), {"token": "aze", "email": emails[0]}}]
return invalid_tokens
```

Why are you defining `tokens` but returning `invalid_tokens` which is undefined? 🤔 Also your schema defines `tokens` field, but your resolver is not returning value for it. Shouldn't it be:

```python
def create_tokens(*_, emails):
    tokens = [{"token": "fsqd", "email": ValueError("boom"), {"token": "aze", "email": emails[0]}}]
    return {"tokens": tokens}
```

?

**Reply by tvallois** [reply]
My test was bad sorry.

Ok your suggestion seems to work as i want:
```python
# Mutation
def create_tokens(*_, emails):
    valid_tokens = [{"data": "aze", "email": email} for email in emails]
    invalid_tokens = [{"data": "fsqd", "email": ValueError("boom")}]
    invalid_tokens.extend(valid_tokens)
    return {"tokens": invalid_tokens}
​
​
mutation_type.set_field("createTokens", create_tokens)

# Test
@pytest.mark.asyncio
async def test_error():
    async with AsyncClient(app=app, base_url="http://localhost") as ac:
        mut = """
        mutation createTokens(
            $emails: [String!]!
        ) {
            createTokens(
                emails: $emails,
            ) {
                tokens {
                    data
                    email
                }
            }
        }
    """
        response = await ac.post(
            "/graphql/",
            json={
                "query": mut,
                "variables": {"emails": ["aze", "eza"]},
            },
        )
        assert response.status_code == 200
        assert response.json() == {'data': {'createTokens': {'tokens': [None, {'data': 'aze', 'email': 'aze'}, {'data': 'aze', 'email': 'eza'}]}}, 'errors': [{'message': 'boom', 'locations': [{'line': 10, 'column': 21}], 'path': ['createTokens', 'tokens', 0, 'email']}]}
```

**Reply by tvallois** [reply]
Thanks for your help !
